crates/oxc_linter/src/rules/jest/no_untyped_mock_factory.rs

```diff
@@ -0,0 +1,410 @@
+use crate::{
+    context::LintContext,
+    fixer::Fix,
+    rule::Rule,
+    utils::{collect_possible_jest_call_node, PossibleJestNode},
+};
+
+use oxc_ast::{
+    ast::{Argument, Expression},
+    AstKind,
+};
+use oxc_diagnostics::{
+    miette::{self, Diagnostic},
+    thiserror::Error,
+};
+use oxc_macros::declare_oxc_lint;
+use oxc_span::Span;
+
+#[derive(Debug, Error, Diagnostic)]
+#[error("eslint-plugin-jest(no-untyped-mock-factory): Disallow using `jest.mock()` factories without an explicit type parameter.")]
+#[diagnostic(
+    severity(warning),
+    help("Add a type parameter to the mock factory such as `typeof import({0:?})`")
+)]
+struct AddTypeParameterToModuleMockDiagnostic(String, #[label] pub Span);
+
+#[derive(Debug, Default, Clone)]
+pub struct NoUntypedMockFactory;
+
+declare_oxc_lint!(
+    /// ### What it does
+    ///
+    /// This rule triggers a warning if `mock()` or `doMock()` is used without a generic
+    /// type parameter or return type.
+    ///
+    /// ### Why is this bad?
+    ///
+    /// By default, `jest.mock` and `jest.doMock` allow any type to be returned by a
+    /// mock factory. A generic type parameter can be used to enforce that the factory
+    /// returns an object with the same shape as the original module, or some other
+    /// strict type. Requiring a type makes it easier to use TypeScript to catch changes
+    /// needed in test mocks when the source module changes.
+    ///
+    /// ### Example
+    ///
+    /// // invalid
+    /// ```typescript
+    /// jest.mock('../moduleName', () => {
+    ///     return jest.fn(() => 42);
+    /// });
+    ///
+    /// jest.mock('./module', () => ({
+    ///     ...jest.requireActual('./module'),
+    ///     foo: jest.fn(),
+    /// }));
+    ///
+    /// jest.mock('random-num', () => {
+    ///     return jest.fn(() => 42);
+    /// });
+    /// ```
+    ///
+    /// // valid
+    /// ```typescript
+    ///
+    /// // Uses typeof import()
+    /// jest.mock<typeof import('../moduleName')>('../moduleName', () => {
+    ///     return jest.fn(() => 42);
+    /// });
+    ///
+    /// jest.mock<typeof import('./module')>('./module', () => ({
+    ///     ...jest.requireActual('./module'),
+    ///     foo: jest.fn(),
+    /// }));
+    ///
+    /// // Uses custom type
+    /// jest.mock<() => number>('random-num', () => {
+    ///     return jest.fn(() => 42);
+    /// });
+    ///
+    /// // No factory
+    /// jest.mock('random-num');
+    ///
+    /// // Virtual mock
+    /// jest.mock(
+    ///     '../moduleName',
+    ///     () => {
+    ///         return jest.fn(() => 42);
+    ///     },
+    ///     { virtual: true },
+    /// );
+    /// ```
+    ///
+    NoUntypedMockFactory,
+    correctness,
+);
+
+impl Rule for NoUntypedMockFactory {
+    fn run_once(&self, ctx: &LintContext<'_>) {
+        for possible_jest_node in &collect_possible_jest_call_node(ctx) {
```

Comment:
Looks like we only run this rule on TS files.

```rust
if !ctx.source_type().is_typescript() {
      return;
}
```

Reply:
Done.